# [1] Library

In [2]:
import pandas as pd

# [2] Data Wrangling

## [2.1] Gathering Data

In [4]:
df = pd.read_csv('./Clean Data/ingredients_idn.csv')
df.head()

nama rating  \
0         3-O Ethyl Ascorbic Acid   Best   
1                            Acai   Best   
2           Acerola Fruit Extract   Best   
3            Acetyl Carnitine HCL   Best   
4  Acetyl Dipeptide-1 Cetyl Ester   Best   

                                         benefit  \
0  Anti-Aging, Evens Skin Tone, Dark Spot Fading   
1                           Anti-Aging, Soothing   
2                                     Anti-Aging   
3                                      Hydration   
4                                     Anti-Aging   

                                            kategori  \
0  Anti-Aging, Evens Skin Tone, Dark Spot Fading,...   
1  Anti-Aging, Soothing, Antioxidant, Plant Extracts   
2            Anti-Aging, Antioxidant, Plant Extracts   
3                             Hydration, Antioxidant   
4                               Anti-Aging, Peptides   

                                          key_points  \
0  A more stable derivative of pure vitamin C | C...   
1  Potent source of skin-beneficial antioxidants ...   
2  A fruit extract also known as Malpighia emargi...   
3                                                NaN   
4  Lab-made anti-aging peptide that promotes firm...   

                                           deskripsi  \
0  3-O ethyl ascorbic acid is a stable, water- an...   
1  Acai (pronounced “ah-sigh-ee”) is a small berr...   
2  Acerola fruit extract (also known as _Malpighi...   
3                                                NaN   
4  Acetyl dipeptide-1 cetyl ester is a lab-engine...   

                                       deskripsi_idn  \
0  Asam askorbat etil 3-O adalah turunan vitamin ...   
1  Acai (diucapkan “ah-sigh-ee”) adalah buah beri...   
2  Ekstrak buah Acerola (juga dikenal sebagai _Ma...   
3                                                NaN   
4  Asetil dipeptida-1 cetil ester adalah peptida ...   

                                         benefit_idn  \
0  Anti Penuaan, Meratakan Warna Kulit, Memudarka...   
1                          Anti Penuaan, Menenangkan   
2                                       Anti Penuaan   
3                                            Hidrasi   
4                                       Anti Penuaan   

                                        kategori_idn  \
0  Anti Penuaan, Meratakan Warna Kulit, Memudarka...   
1  Anti Penuaan, Menenangkan, Antioksidan, Ekstra...   
2        Anti Penuaan, Antioksidan, Ekstrak Tumbuhan   
3                               Hidrasi, Antioksidan   
4                              Anti Penuaan, Peptida   

                                      key_points_idn  
0  Turunan vitamin C murni yang lebih stabil | Ma...  
1  Sumber antioksidan yang ampuh untuk kulit | Me...  
2  Ekstrak buah yang juga dikenal sebagai Malpigh...  
3                                                NaN  
4  Peptida anti-penuaan buatan laboratorium yang ...

Insight:
1. Terdapat 10 kolom yaitu nama, rating, benefit, kategori, key_points, deskripsi, deskripsi_idn, benefit_ind, kategori_idn, dan key_points_idn
2. Kolom rating, benefit, kategori, key_points, dan deskripsi merupakan bahasa inggris, tidak digunakan
3. Kolom deskripsi_idn, benefit_idn, kategori_idn, dan key_points_idn merupakan bahasa indonesia, digunakan

Note:
1. Drop semua kolom berbahasa inggris

## [2.2] Assessing Data

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26043 entries, 0 to 26042
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   nama            26043 non-null  object
 1   rating          26043 non-null  object
 2   benefit         1546 non-null   object
 3   kategori        2389 non-null   object
 4   key_points      1160 non-null   object
 5   deskripsi       26014 non-null  object
 6   deskripsi_idn   26014 non-null  object
 7   benefit_idn     1546 non-null   object
 8   kategori_idn    2389 non-null   object
 9   key_points_idn  1160 non-null   object
dtypes: object(10)
memory usage: 2.0+ MB


Insight:
1. Kolom benefit, kategori, key_points, benefit_idn, kategori_idn, key_points_idn adalah kolom dengan jumlah data tersedikit. Hal ini disebabkan oleh keterbatasan data dari sumber
2. Terdapat 26043 ingredients dan semua telah memiliki nilai pada kolom rating walau tidak semuanya memiliki deskripsi

Note:
1. Kolom dengan jumlah data sedikit tidak perlu didrop atau dibuat nilai baru karena keterbatasan data dari sumber data

In [6]:
df.isna().sum()

nama                  0
rating                0
benefit           24497
kategori          23654
key_points        24883
deskripsi            29
deskripsi_idn        29
benefit_idn       24497
kategori_idn      23654
key_points_idn    24883
dtype: int64

Insight:
1. Jumlah null untuk benefit, kategori, key_points, benefit_idn, kategori_idn, dan key_points_idn sangat tinggi yaitu lebih dari 23 ribu
2. Jumlah null untuk deskripsi sedikit yaitu hanya 29

Note:
1. Walaupun jumlah poin nomor 1 sangat banyak null, kolom tidak bisa didrop karena informasi di kolom tersebut sangat penting. Apabila mencari 23 ribu data lainnya, maka akan cukup sulit karena keterbatasan sumber informasi
2. Begitu pula untuk poin nomor 2, karena keterbatasan sumber informasi, maka tidak bisa menambahkan atau mencari data manual

In [7]:
df.duplicated().sum()

7

Insight:
1. Terdapat 7 duplikat

Note:
1. Hapus semua duplikat

In [8]:
df.describe()

nama     rating    benefit  \
count                             26043      26043       1546   
unique                            26031          6         83   
top     Borago Officinalis Seed Extract  Not Rated  Hydration   
freq                                  3      23626        512   

                    kategori  \
count                   2389   
unique                   666   
top     Hydration, Humectant   
freq                      88   

                                               key_points  \
count                                                1160   
unique                                               1147   
top     Rich in antioxidants that defend skin against ...   
freq                                                    2   

                                                deskripsi  \
count                                               26014   
unique                                               4160   
top     Functions: Skin Conditioning. We have not yet ...   
freq                                                 5624   

                                            deskripsi_idn benefit_idn  \
count                                               26014        1546   
unique                                               4149          83   
top     Fungsi: Merawat Kulit. Kami belum menilai baha...     Hidrasi   
freq                                                 5624         512   

             kategori_idn                                     key_points_idn  
count                2389                                               1160  
unique                666                                               1147  
top     Hidrasi, Humektan  Kaya akan antioksidan yang melindungi kulit da...  
freq                   88                                                  2

Insight:
1. Terdapat 26043 nama, 26043 rating, 1546 benefit, 2389 kategori, 1160 key_points, dan 26014 deskripsi (begitu juga untuk kolom translate bahasa indonesia)
2. Terdapat 26031 nama unique, 6 unique rating, 83 unique benefit, 666 unique kategori, 1147 unique key_points, dan 4160 unique deskripsi (begitu juga untuk kolom translate bahasa indonesia)


Note:
1. Semua nama haruslah unique, jika tidak berarti ada nama yang sama. Drop

In [9]:
print(df['nama'])

0                           3-O Ethyl Ascorbic Acid
1                                              Acai
2                             Acerola Fruit Extract
3                              Acetyl Carnitine HCL
4                    Acetyl Dipeptide-1 Cetyl Ester
                            ...                    
26038                        Zostera Marina Extract
26039       Zygophyllum Qatarense Leaf/Stem Extract
26040    Zygosaccharomyces Microellipsoides Ferment
26041                     Zymomonas Ferment Extract
26042                    Zymomonas Ferment Filtrate
Name: nama, Length: 26043, dtype: object


Insight:
1. Dapat dilihat terdapat data yang menggunakan /. Untuk memudahkan pencocokan, maka akan dipisah menjadi 2 row baris berbeda

In [16]:
print(df['benefit_idn'])

0        Anti Penuaan, Meratakan Warna Kulit, Memudarka...
1                                Anti Penuaan, Menenangkan
2                                             Anti Penuaan
3                                                  Hidrasi
4                                             Anti Penuaan
                               ...                        
26038                                                  NaN
26039                                                  NaN
26040                                                  NaN
26041                                                  NaN
26042                                                  NaN
Name: benefit_idn, Length: 26043, dtype: object


In [17]:
print(df['kategori_idn'])

0        Anti Penuaan, Meratakan Warna Kulit, Memudarka...
1        Anti Penuaan, Menenangkan, Antioksidan, Ekstra...
2              Anti Penuaan, Antioksidan, Ekstrak Tumbuhan
3                                     Hidrasi, Antioksidan
4                                    Anti Penuaan, Peptida
                               ...                        
26038                                                  NaN
26039                                                  NaN
26040                                                  NaN
26041                                                  NaN
26042                                                  NaN
Name: kategori_idn, Length: 26043, dtype: object


In [18]:
print(df['key_points_idn'])

0        Turunan vitamin C murni yang lebih stabil | Ma...
1        Sumber antioksidan yang ampuh untuk kulit | Me...
2        Ekstrak buah yang juga dikenal sebagai Malpigh...
3                                                      NaN
4        Peptida anti-penuaan buatan laboratorium yang ...
                               ...                        
26038                                                  NaN
26039                                                  NaN
26040                                                  NaN
26041                                                  NaN
26042                                                  NaN
Name: key_points_idn, Length: 26043, dtype: object


In [20]:
print(df['deskripsi_idn'])

0        Asam askorbat etil 3-O adalah turunan vitamin ...
1        Acai (diucapkan “ah-sigh-ee”) adalah buah beri...
2        Ekstrak buah Acerola (juga dikenal sebagai _Ma...
3                                                      NaN
4        Asetil dipeptida-1 cetil ester adalah peptida ...
                               ...                        
26038    Fungsi: Merawat Kulit. Kami belum menilai baha...
26039    Fungsi: Antioksidan. Kami belum menilai bahan ...
26040    Fungsi: Humektan. Kami belum menilai bahan ini...
26041    Fungsi: Humektan, Pengkondisi Kulit. Kami belu...
26042    Fungsi: Humektan, Pengkondisi Kulit. Kami belu...
Name: deskripsi_idn, Length: 26043, dtype: object


In [14]:
unique_rating = df['rating'].unique()
print(unique_rating)

['Best' 'Good' 'Average' 'Bad' 'Worst' 'Not Rated']


Insight:
1. Terdapat 6 rating berbeda yaitu Best, Good, Average, Bad, Worst, Not Rated

Note:
1. Ubah menjadi bahasa indonesia

In [21]:
unique_kategori = df['kategori_idn'].str.split(',').explode().unique()
print(unique_kategori)

['Anti Penuaan' ' Meratakan Warna Kulit' ' Memudarkan Flek Hitam'
 ' Antioksidan' ' Menenangkan' ' Ekstrak Tumbuhan' 'Hidrasi' ' Peptida'
 'Melembabkan' ' Humektan' ' Hidrasi' 'Antioksidan' ' Agen Chelating'
 ' Anti Jerawat' ' Emollient' 'Ekstrak Tumbuhan yang Menenangkan'
 'Menenangkan' ' Anti Penuaan' ' Melembabkan' ' Mengelupas'
 ' Penambah Tekstur' ' Emolien' ' Agen Khelasi'
 ' Prebiotik/Probiotik/Postbiotik' 'Meratakan Warna Kulit' ' Eksfoliator'
 'Memudarkan Flek Hitam' ' Pengatur/Penstabil pH' ' Pewarna/Pigmen'
 ' Filter UV' 'Anti-jerawat' ' Memudarkan flek hitam'
 ' Menghaluskan kulit yang berjerawat' 'Penenang'
 ' menghaluskan kulit yang berbenjol-benjol' ' antibakteri'
 ' Mengurangi komedo' 'Pelembab' ' Pelembab' nan ' Pengontrol Minyak'
 ' Pengemulsi' ' Penguat Tekstur' ' Agen Pembersih'
 ' Agen Oklusif/Pengaburan' 'Yg melunakkan' ' Pelarut' ' Silikon'
 'Emolien' 'Filter UV' 'Ekstrak Tumbuhan' 'Agen hidrasi' ' emolien'
 ' oklusif/pengaburan' ' Agen Pembentuk Lapisan' ' Emuls

In [23]:
unique_key_points = df['key_points_idn'].str.split('|').explode().unique()
print(unique_key_points)

['Turunan vitamin C murni yang lebih stabil '
 ' Mampu membantu memudarkan bintik hitam '
 ' Memiliki sifat yang larut dalam air dan minyak ' ...
 ' Dikatakan memiliki sifat pengawet dan anti-jerawat '
 ' Laporan dan penelitian yang saling bertentangan seputar kemampuannya untuk membantu kulit yang rentan berjerawat '
 ' Dapat menyebabkan sensitivitas kulit jika dioleskan']


Kesimpulan:
1. Drop semua kolom berbahasa inggris
2. Drop semua duplikat
3. Ubah rating menjadi bahasa indonesia dengan mapping agar lebih cepat
4. Apabila ada penggunaan / pada nama, maka pisahkan menjadi baris sendiri untuk setiap namanya
5. (PENGECEKAN MANUAL) Ditemukan beberapa karakter khusus, perbaiki

# [2.3] Cleaning Data

###  [2.3.1] Drop semua kolom berbahasa inggris

In [24]:
delete = ['benefit', 'kategori' ,'key_points', 'deskripsi']
df = df.drop(columns=delete)

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26043 entries, 0 to 26042
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   nama            26043 non-null  object
 1   rating          26043 non-null  object
 2   deskripsi_idn   26014 non-null  object
 3   benefit_idn     1546 non-null   object
 4   kategori_idn    2389 non-null   object
 5   key_points_idn  1160 non-null   object
dtypes: object(6)
memory usage: 1.2+ MB


Insight:
1. Kolom benefit, kategori, key_points, dan deskripsi telah berhasil terhapus

### [2.3.2] Drop semua duplikat

In [26]:
nama_duplikat = df['nama'][df['nama'].duplicated()].unique()
print(nama_duplikat)

['AHA' 'BHA (Beta Hydroxy Acid)' 'Borago Officinalis Seed Extract'
 'Fucus Vesiculosus Extract' 'Persea Gratissima (Avocado) Oil'
 'Rubus Idaeus (Raspberry) Fruit Extract'
 'Rubus Occidentalis Fruit Extract' 'Skin-Replenishing Ingredients'
 'Pterocarpus Santalinus' 'Advanced Glycation Endproduct (AGE)' 'Albumen']


In [27]:
df = df.drop_duplicates(subset=['nama'], keep='first')

In [28]:
nama_duplikat = df['nama'][df['nama'].duplicated()].unique()
print(nama_duplikat)

[]


Insight:
1. Semua duplikat berhasil terhapus

### [2.3.3] Ubah rating menjadi bahasa indonesia dengan mapping

In [ ]:
rating_mapping = {
    'Best': 'Terbaik',
    'Good': 'Baik',
    'Average': 'Rata-Rata',
    'Bad': 'Buruk',
    'Worst': 'Terburuk',
    'Not Rated': 'Belum Dinilai'}

df['rating'] = df['rating'].replace(rating_mapping)

In [31]:
unique_rating = df['rating'].unique()
print(unique_rating)

['Terbaik' 'Baik' 'Rata-Rata' 'Buruk' 'Terburuk' 'Belum Dinilai']


Insight:
1. Rating telah diubah menjadi bahasa indonesia

### [2.3.4] Memecah nama / menjadi satu untuk setiap baris

In [ ]:
def split_and_duplicate(row):
    if '/' in row['nama']:
        names = row['nama'].split('/')
        duplicated_rows = pd.DataFrame(
            {col: [row[col]] * len(names) for col in row.index}
        )
        duplicated_rows['nama'] = names
        return duplicated_rows
    else:
        return pd.DataFrame([row])

df_split = pd.concat(split_and_duplicate(row) for _, row in df.iterrows()).reset_index(drop=True)

print(df_split['nama'])

0                           3-O Ethyl Ascorbic Acid
1                                              Acai
2                             Acerola Fruit Extract
3                              Acetyl Carnitine HCL
4                    Acetyl Dipeptide-1 Cetyl Ester
                            ...                    
28139                    Zygophyllum Qatarense Leaf
28140                                  Stem Extract
28141    Zygosaccharomyces Microellipsoides Ferment
28142                     Zymomonas Ferment Extract
28143                    Zymomonas Ferment Filtrate
Name: nama, Length: 28144, dtype: object


In [37]:
has_slash = df_split['nama'].str.contains('/').any()
print(has_slash)

False


Insight:
1. Sudah tidak ada nama yang menggunakan / dibuktikan dengan has_slash yang false dan jumlah data pada kolom nama yang semula 26 ribu sekarang menjadi 28 ribu

In [38]:
df_split.to_csv('./Clean Data/ingredients.csv', index=False)